tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_replay_cache.c

```diff
@@ -0,0 +1,203 @@
+	.addr = 0x0001,
+	.dev_key = { 0x01 },
+};
+static const struct bt_mesh_test_cfg rx_cfg = {
+	.addr = 0x0002,
+	.dev_key = { 0x02 },
+};
+
+static bool are_settings_kept;
+static bool is_tx_succeeded;
+
+static void test_tx_init(void)
+{
+	bt_mesh_test_cfg_set(&tx_cfg, WAIT_TIME);
+}
+
+static void test_rx_init(void)
+{
+	bt_mesh_test_cfg_set(&rx_cfg, WAIT_TIME);
+}
+
+static void test_delete(void)
+{
+	if (!are_settings_kept) {
+		settings_test_backend_clear();
+	}
+}
+
+static void tx_started(uint16_t dur, int err, void *data)
+{
+	if (err) {
+		FAIL("Couldn't start sending (err: %d)", err);
+	}
+
+	LOG_INF("Sending started");
+}
+
+static void tx_ended(int err, void *data)
+{
+	struct k_sem *sem = data;
+
+	if (err) {
+		is_tx_succeeded = false;
+		LOG_INF("Sending failed (%d)", err);
+	} else {
+		is_tx_succeeded = true;
+		LOG_INF("Sending succeeded");
+	}
+
+	k_sem_give(sem);
+}
+
+static void test_tx_immediate_replay_attack(void)
+{
+	bt_mesh_test_setup();
+
+	static const struct bt_mesh_send_cb send_cb = {
+		.start = tx_started,
+		.end = tx_ended,
+	};
+	struct k_sem sem;
+
+	k_sem_init(&sem, 0, 1);
+
+	uint32_t seq = bt_mesh.seq;
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(bt_mesh_test_send_ra(rx_cfg.addr, 20, &send_cb, &sem));
+
+		if (k_sem_take(&sem, K_SECONDS(TEST_DATA_WAITING_TIME))) {
+			LOG_ERR("Send timed out");
+		}
+
+		ASSERT_TRUE(is_tx_succeeded);
+	}
+
+	bt_mesh.seq = seq;
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(bt_mesh_test_send_ra(rx_cfg.addr, 20, &send_cb, &sem));
+
+		if (k_sem_take(&sem, K_SECONDS(TEST_DATA_WAITING_TIME))) {
+			LOG_ERR("Send timed out");
+		}
+
+		ASSERT_TRUE(!is_tx_succeeded);
+	}
+
+	PASS();
+}
+
+static void test_rx_immediate_replay_attack(void)
+{
+	bt_mesh_test_setup();
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(bt_mesh_test_recv(20, rx_cfg.addr,
+				K_SECONDS(TEST_DATA_WAITING_TIME)), "RX fail");
+	}
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(!bt_mesh_test_recv(20, rx_cfg.addr,
+				K_SECONDS(TEST_DATA_WAITING_TIME)), "RX fail");
+	}
+
+	are_settings_kept = true;
+
+	PASS();
+}
+
+static void test_tx_power_replay_attack(void)
+{
+	bt_mesh_test_setup();
+
+	static const struct bt_mesh_send_cb send_cb = {
+		.start = tx_started,
+		.end = tx_ended,
+	};
+	struct k_sem sem;
+
+	k_sem_init(&sem, 0, 1);
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(bt_mesh_test_send_ra(rx_cfg.addr, 20, &send_cb, &sem));
+
+		if (k_sem_take(&sem, K_SECONDS(TEST_DATA_WAITING_TIME))) {
+			LOG_ERR("Send timed out");
+		}
+
+		ASSERT_TRUE(!is_tx_succeeded);
+	}
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(bt_mesh_test_send_ra(rx_cfg.addr, 20, &send_cb, &sem));
+
+		if (k_sem_take(&sem, K_SECONDS(TEST_DATA_WAITING_TIME))) {
+			LOG_ERR("Send timed out");
+		}
+
+		ASSERT_TRUE(is_tx_succeeded);
+	}
+
+	PASS();
+}
+
+static void test_rx_power_replay_attack(void)
+{
+	bt_mesh_test_setup();
+
+	for (int i = 0; i < 3; i++) {
+		ASSERT_OK(bt_mesh_test_recv(20, rx_cfg.addr,
```

Comment:
IMO it's worth to encode something meaningful to differentiate messages, e.g. message number. Thus, on the receiving side we could actually check that only the first 3 messages (in case of immediate replay attack test) and only the last 3 messages (after booting up the receiving side) were received. Current approach just checks that some 3 messages were received.

Reply:
added payload analysis. Generally, it is not correct to perform replay attack by frame with a different payload. The only sequence of succeeded and failed frames and accounting their number can give the ability to make decision about working or broken replay protection cache.